File: 백준/Silver/1189. 컴백홈/컴백홈.cc
#include<bits/stdc++.h>
using namespace std;
const int dy[]={-1,0,1,0};
const int dx[]={0,1,0,-1};
int r,c,k,ny,nx,ret;
char a[7][7];
int visited[7][7];
void DFS(int y,int x,int cnt)
{
	if(y==0 && x==c-1)
	{
		if(cnt==k) ret++;
		return;
	}
	visited[y][x]=1;
	for(int i=0;i<4;i++)
	{
		ny=y+dy[i];
		nx=x+dx[i];
		if(ny<0 || ny>=r || nx<0 || nx>=c || visited[ny][nx] || a[ny][nx]=='T') continue;
		DFS(ny,nx,cnt+1);
	}
	visited[y][x]=0;
}
int main()
{
	cin>>r>>c>>k;
	for(int i=0;i<r;i++)
	{
		for(int j=0;j<c;j++)
		{
			cin>>a[i][j];
		}
	}
	DFS(r-1,0,1);
	cout<<ret;
}

File: 백준/Silver/1189. 컴백홈/컴백홈.py
#import sys
#sys.setrecursionlimit(10000)
dy=[-1,0,1,0]
dx=[0,1,0,-1]
ret=0
def dfs(y,x,cnt):
    global ret
    if cnt==k:
        if y==0 and x==c-1:
            ret+=1
        return

    for i in range(4):
        ny = y+dy[i]
        nx = x+dx[i]
        if ny<0 or ny>=r or nx<0 or nx>=c or visited[ny][nx] or a[ny][nx]=='T':
            continue
        if not visited[ny][nx] and a[ny][nx]=='.':
            visited[ny][nx]=True
            dfs(ny, nx, cnt + 1)
            visited[ny][nx]=False


r,c,k = map(int,input().split())

a=[['']*c for _ in range(r)]
visited=[[0]*c for _ in range(r)]

for i in range(r):
    line=input().strip()
    for j in range(c):
        a[i][j]=line[j]

visited[r-1][0]=True
dfs(r-1,0,1)
print(ret)